### PR TITLE
New instructor button to remove all students from exam

### PR DESCRIPTION
**Pull request by nicolasruscher**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

Currently when instructors want to remove all students from an exam they have to remove them all by hand. In this PR we fix this.

### Description

Solve https://github.com/ls1intum/Artemis/issues/2935, by introducing a new button to remove all students at once. 

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration -> Exams -> students
3. test the two options: removing all students from the exam with the check box checked will delete all participations and submissions as well. 
4. Without the checkout, the participations and submissions should still exist, and all students only go unregistered from the exam without deleting existing stuff. 

### Test Coverage

Client:
exam-students.component: 83.33%

Server:
ExamResource: 95%
ExamRegistrationService: 94%

### Screenshots

Bildschirmfoto 2021-03-13 um 00 41 17

Bildschirmfoto 2021-03-08 um 17 35 43

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java [hunk 1]
```diff
@@ -683,6 +688,32 @@ public ExamResource(UserRepository userRepository, CourseRepository courseReposi
...
+     * Remove all students of the exam so that they cannot access the exam any more.
+     * Optionally, also deletes participations and submissions of all students in their student exams.
+     *
+     * @param courseId     the id of the course
+     * @param examId       the id of the exam
+     * @param withParticipationsAndSubmission request param deciding whether participations and submissions should also be deleted
+     *
+     * @return empty ResponseEntity with status 200 (OK) or with status 404 (Not Found)
+     */
+    @DeleteMapping(value = "/courses/{courseId}/exams/{examId}/allstudents")
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java:702]: In terms of REST, we would want to delete the Collection:
```suggestion
    @DeleteMapping(value = "/courses/{courseId}/exams/{examId}/students")
```

**Review by jpbernius — COMMENTED** [review]
This looks already very good.
On the code:
- Codacy has some valid ideas

**Review by tiffca — DISMISSED** [review]
Works as described

File: src/main/webapp/app/exam/manage/students/exam-students.component.html [hunk 2]
```diff
@@ -24,6 +24,25 @@ <h3 class="mb-0">
                 <fa-icon [icon]="'plus'" [fixedWidth]="true"></fa-icon>&nbsp;
                 <span jhiTranslate="artemisApp.examManagement.examStudents.registerAllFromCourse"></span>
             </button>
+            <button
```
Comment (SimonEntholzer) [inline src/main/webapp/app/exam/manage/students/exam-students.component.html:27]: I think this button should get a similar size as the other buttons.
image

File: src/main/webapp/i18n/de/exam.json [hunk 3]
```diff
@@ -189,6 +189,11 @@
             "unlockAllRepositoriesFailure": "Es gab einen Fehler beim Entsperren der Repositories:\n {{message}}",
             "lockAllRepositoriesSuccess": "Repositories von {{number}} Programmieraufgaben wurden gesperrt.",
             "lockAllRepositoriesFailure": "Es gab einen Fehler beim Sperren der Repositories:\n {{message}}",
+            "removeAllStudents": {
+                "title": "Entferne alle Studierenden",
+                "question": "Bist du sicher, dass du <strong class='text-danger'>alle Studierenden</strong> von der Klausur abmelden möchtest?",
+                "confirmationText": "Bitte gibt den Namen der Klausur ein, um das Entfernen aller Studierenden von der Klausur zu bestätigen!!"
```
Comment (SimonEntholzer) [inline src/main/webapp/i18n/de/exam.json:195]:
```suggestion
                "confirmationText": "Bitte gib den Namen der Klausur ein, um das Entfernen aller Studierenden von der Klausur zu bestätigen!!"
```

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]
The code looks good, and removing the students in the intended way works too. I only have two remarks:
1. In my opinion the button should have the same size as the others.
2. Something I noticed while testing:
I created an exam, added students, participated with one, also handed in. Now, I unregistered all students from the exam, but did not check the checkbox. -> so far so good, I could still see the submission of the student in the assessment dashboards. 
Now, after registering all students again, there is kind of an issue. Since the participation/submission of the student are orphaned, as no student exam for them exists anymore, the student is not able to see results provided by the instructor. 
This also means, that clicking the "remove all students" button with submissions/participations checked does not delete those, as they are not connected to any student exam. 
Maybe we should think of a way to handle this. (a way to recover/recreate the deleted studentExams?)

File: src/main/webapp/app/exam/manage/students/exam-students.component.html (hunk 2 above)
Comment (nicolasruscher) [reply src/main/webapp/app/exam/manage/students/exam-students.component.html:27]: Thanks for catching that! Done 👍

**Comment by nicolasruscher** [comment]
> The code looks good, and removing the students in the intended way works too. I only have two remarks:
> 
> 1. In my opinion the button should have the same size as the others.
> 2. Something I noticed while testing:
>    I created an exam, added students, participated with one, also handed in. Now, I unregistered all students from the exam, but did not check the checkbox. -> so far so good, I could still see the submission of the student in the assessment dashboards.
>    Now, after registering all students again, there is kind of an issue. Since the participation/submission of the student are orphaned, as no student exam for them exists anymore, the student is not able to see results provided by the instructor.
>    This also means, that clicking the "remove all students" button with submissions/participations checked does not delete those, as they are not connected to any student exam.
>    Maybe we should think of a way to handle this. (a way to recover/recreate the deleted studentExams?)

I think the current behavior is expected as it is the same when deleting a single user and then registering him again. If this is a proper concern it should probably be addressed for the single user deletion as well, and therefore probably in a follow up.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java (hunk 1 above)
Comment (nicolasruscher) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java]: Thanks for the feedback, done 👍

**Review by SimonEntholzer — APPROVED** [review]
Tested again, now the button looks nice :D

File: src/main/java/de/tum/in/www1/artemis/config/Constants.java [hunk 4]
```diff
@@ -128,6 +128,8 @@
 
     public static final String REMOVE_USER_FROM_EXAM = "REMOVE_USER_FROM_EXAM";
 
+    public static final String REMOVE_ALL_USERS_FROM_EXAM = "REMOVE_ALL_USERs_FROM_EXAM";
```
Comment (nicolasruscher) [inline src/main/java/de/tum/in/www1/artemis/config/Constants.java:131]:
```suggestion
    public static final String REMOVE_ALL_USERS_FROM_EXAM = "REMOVE_ALL_USERS_FROM_EXAM";
```

**Review by ivanchimeno — APPROVED** [review]
Code looks good!

File: src/main/webapp/i18n/de/exam.json [hunk 5]
```diff
@@ -189,6 +189,11 @@
             "unlockAllRepositoriesFailure": "Es gab einen Fehler beim Entsperren der Repositories:\n {{message}}",
             "lockAllRepositoriesSuccess": "Repositories von {{number}} Programmieraufgaben wurden gesperrt.",
             "lockAllRepositoriesFailure": "Es gab einen Fehler beim Sperren der Repositories:\n {{message}}",
+            "removeAllStudents": {
+                "title": "Entferne alle Studierenden",
+                "question": "Bist du sicher, dass du <strong class='text-danger'>alle Studierenden</strong> von der Klausur abmelden möchtest?",
+                "confirmationText": "Bitte gib den Namen der Klausur ein, um das Entfernen aller Studierenden von der Klausur zu bestätigen!!"
```
Comment (nicolasruscher) [inline src/main/webapp/i18n/de/exam.json:195]:
```suggestion
                "confirmationText": "Bitte gib den Namen der Klausur ein, um das Entfernen aller Studierenden von der Klausur zu bestätigen!"
```

**Review by SimonEntholzer — APPROVED** [review]
changes make sense :D re-approve

**Review by ivohashamov — APPROVED** [review]
Tested successfully on ts1

**Review by iozbeyli — APPROVED** [review]
Code looks good :+1:

**Pull request merged**